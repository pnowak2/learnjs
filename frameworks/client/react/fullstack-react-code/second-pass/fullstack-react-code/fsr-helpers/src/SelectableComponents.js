import React from "react";

export class ComponentMenu extends React.Component {
  state = {
    visible: false
  };
  handleClick = () => {
    this.setState(prevState => ({
      visible: !prevState.visible
    }));
  };

  render() {
    const activeClass = this.state.visible ? "simple active" : "";
    return (
      <div
        className="ui text menu"
        style={{ marginRight: "50px" }}
        onClick={this.handleClick}
      >
        <div className={`ui right dropdown item ${activeClass}`}>
          Select Component
          <i className="dropdown icon" />
          <div className={`menu`}>
            {this.props.slugs.map((slug, idx) => {
              const divClass = (slug === this.props.activeSlug) ? 'active item': 'item'
              return (
                <div
                  className={divClass}
                  id={slug} // for nightwatch
                  key={slug}
                  onClick={() => this.props.onComponentSelect(slug)}
                >
                  {slug}
                </div>
              );
            })}
          </div>
        </div>
      </div>
    );
  }
}

class SelectableComponents extends React.Component {
  constructor(props) {
    super(props);

    let activeSlug = Object.keys(this.props.components)[0];

    const search = (window && window.location.search) || this.props.search;

    if (search) {
      const params = getUrlParams(search);
      if (params.app_slug) {
        activeSlug = params.app_slug;
      }
    }

    this.state = {
      activeSlug
    };
  }
  handleComponentSelect = slug => {
    this.setState({
      activeSlug: slug
    });
  };

  render() {
    const { components } = this.props;
    const component = components[this.state.activeSlug];
    const slugs = Object.keys(components);

    return (
      <div>
        <ComponentMenu
          onComponentSelect={this.handleComponentSelect}
          activeSlug={this.state.activeSlug}
          slugs={slugs}
        />
        {React.createElement(component)}
      </div>
    );
  }
}

function getUrlParams(search) {
  const hashes = search.slice(search.indexOf("?") + 1).split("&");
  const params = {};
  hashes.map(hash => {
    const [key, val] = hash.split("=");
    params[key] = val;
  });

  return params;
}

export default SelectableComponents;
