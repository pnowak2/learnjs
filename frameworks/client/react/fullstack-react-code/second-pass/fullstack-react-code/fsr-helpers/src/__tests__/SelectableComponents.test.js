import React from "react";
import { shallow } from "enzyme";

import SelectableComponents, { ComponentMenu } from "../SelectableComponents";

const App1 = () => {};
const App2 = () => {};

const components = {
  "App-1": App1,
  "App-2": App2
};

describe("SelectableComponents", () => {
  let wrapper;

  describe("selecting via UI", () => {
    beforeEach(() => {
      wrapper = shallow(<SelectableComponents components={components} />);
    });

    it("defaults to rendering first component", () => {
      expect(wrapper.containsMatchingElement(<App1 />)).toBe(true);
    });

    it("renders `ComponentMenu` with the right props", () => {
      const el = wrapper.find("ComponentMenu");
      expect(el.props()).toMatchSnapshot();
    });

    describe("selecting another component", () => {
      beforeEach(() => {
        wrapper.instance().handleComponentSelect("App-2");
        wrapper.update();
      });

      it("should update the state", () => {
        expect(wrapper.state()).toMatchSnapshot();
      });

      it("should render the second component", () => {
        expect(wrapper.containsMatchingElement(<App2 />)).toBe(true);
      });
    });
  });

  describe("selecting via query params", () => {
    const props = {
      components
    };

    it("should select component on mount based on query params", () => {
      props.search = "?app_slug=App-1";
      wrapper = shallow(<SelectableComponents {...props} />);
      expect(wrapper.state()).toMatchSnapshot();
      props.search = "?app_slug=App-2";
      wrapper = shallow(<SelectableComponents {...props} />);
      expect(wrapper.state()).toMatchSnapshot();
    });
  });
});

describe("ComponentMenu", () => {
  let wrapper;
  const onComponentSelect = jest.fn();
  const slugs = Object.keys(components);

  beforeEach(() => {
    wrapper = shallow(
      <ComponentMenu onComponentSelect={onComponentSelect} slugs={slugs} />
    );
  });

  it("renders a list of components", () => {
    expect(
      wrapper.find("div > .menu > .item").map(i => i.html())
    ).toMatchSnapshot();
  });

  it("calls `onComponentSelect` when one is clicked", () => {
    wrapper.find("#App-1").simulate("click");
    expect(onComponentSelect.mock.calls.length).toEqual(1);
    let callArgs = onComponentSelect.mock.calls[0];
    expect(callArgs[0]).toEqual("App-1");

    wrapper.find("#App-2").simulate("click");
    expect(onComponentSelect.mock.calls.length).toEqual(2);
    callArgs = onComponentSelect.mock.calls[1];
    expect(callArgs[0]).toEqual("App-2");
  });
});
