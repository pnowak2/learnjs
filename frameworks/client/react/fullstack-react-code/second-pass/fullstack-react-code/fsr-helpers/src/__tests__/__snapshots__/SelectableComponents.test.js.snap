// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentMenu renders a list of components 1`] = `
Array [
  "<div class=\\"item\\" id=\\"App-1\\">App-1</div>",
  "<div class=\\"item\\" id=\\"App-2\\">App-2</div>",
]
`;

exports[`SelectableComponents selecting via UI renders \`ComponentMenu\` with the right props 1`] = `
Object {
  "activeSlug": "App-1",
  "onComponentSelect": [Function],
  "slugs": Array [
    "App-1",
    "App-2",
  ],
}
`;

exports[`SelectableComponents selecting via UI selecting another component should update the state 1`] = `
Object {
  "activeSlug": "App-2",
}
`;

exports[`SelectableComponents selecting via query params should select component on mount based on query params 1`] = `
Object {
  "activeSlug": "App-1",
}
`;

exports[`SelectableComponents selecting via query params should select component on mount based on query params 2`] = `
Object {
  "activeSlug": "App-2",
}
`;
